components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/tfm_impl/t_cose/src/t_cose_util.c

```diff
@@ -0,0 +1,189 @@
+    QCBORError                  qcbor_result;
+    struct t_cose_crypto_hash   hash_ctx;
+    int32_t                     hash_alg_id;
+
+    /* This builds the CBOR-format to-be-signed bytes */
+    QCBOREncode_Init(&cbor_encode_ctx, buffer_for_TBS_first_part);
+    QCBOREncode_OpenArray(&cbor_encode_ctx);
+    /* context */
+    QCBOREncode_AddSZString(&cbor_encode_ctx,
+                            COSE_SIG_CONTEXT_STRING_SIGNATURE1);
+    /* body_protected */
+    QCBOREncode_AddBytes(&cbor_encode_ctx,
+                         protected_headers);
+    /* sign_protected */
+    QCBOREncode_AddBytes(&cbor_encode_ctx, NULL_USEFUL_BUF_C);
+    /* external_aad */
+    QCBOREncode_AddBytes(&cbor_encode_ctx, NULL_USEFUL_BUF_C);
+    /* fake payload */
+    QCBOREncode_AddBytes(&cbor_encode_ctx, NULL_USEFUL_BUF_C);
+    QCBOREncode_CloseArray(&cbor_encode_ctx);
+
+    /* get the result and convert it to struct useful_buf_c representation */
+    qcbor_result = QCBOREncode_Finish(&cbor_encode_ctx, &tbs_first_part);
+    if(qcbor_result) {
+        /* Mainly means that the protected_headers were too big
+         (which should never happen) */
+        return_value = T_COSE_ERR_SIG_STRUCT;
+        goto Done;
+    }
+
+    /* Start the hashing */
+    hash_alg_id = hash_alg_id_from_sig_alg_id(cose_alg_id);
+    /* Don't check hash_alg_id for failure. t_cose_crypto_hash_start()
+     will handle it properly
+     */
+    return_value = t_cose_crypto_hash_start(&hash_ctx, hash_alg_id);
+    if(return_value) {
+        goto Done;
+    }
+
+    /* Hash the first part of the TBS. Take all but the last two
+     * bytes. The last two bytes are the fake payload from above. It
+     * is replaced by the real payload which is hashed next. The fake
+     * payload is needed so the array count is right. This is one of
+     * the main things that make it possible to implement with one
+     * buffer for the whole cose sign1.
+     */
+    t_cose_crypto_hash_update(&hash_ctx,
+                              useful_buf_head(tbs_first_part,
+                                              tbs_first_part.len - 2));
+
+    /* Hash the payload */
+    t_cose_crypto_hash_update(&hash_ctx, payload);
+
+    /* Finish the hash and set up to return it */
+    return_value = t_cose_crypto_hash_finish(&hash_ctx,
+                                             buffer_for_hash,
+                                             hash);
+
+Done:
+    return return_value;
+}
+
+
+/*
+ * Public function. See t_cose_util.h
+ */
+enum t_cose_err_t
+get_short_circuit_kid(struct useful_buf buffer_for_kid,
+                     struct useful_buf_c *kid)
+{
+    /* This is a random hard coded key ID that is used to indicate
+     * short-circuit signing. It is OK to hard code this as the
+     * probability of collision with this ID is very low and the same
+     * as for collision between any two key IDs of any sort.
+     */
+    uint8_t defined_short_circuit_kid[] = {
```

Comment:
This should be `static const`, presumably?

Reply:
This file is taken from TFM attestation code.
Should we review this code too?

Reply:
If it's upstream code, may be best to just leave it if making changes causes admin hassle.